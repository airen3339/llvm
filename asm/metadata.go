package asm

import (
	"github.com/llir/ll/ast"
	"github.com/llir/llvm/ir"
)

// irMetadataAttachments returns the IR metadata attachments corresponding to
// the given AST metadata attachments.
func (gen *generator) irMetadataAttachments(ns []ast.MetadataAttachment) []*ir.MetadataAttachment {
	var mds []*ir.MetadataAttachment
	for _, n := range ns {
		md := gen.irMetadataAttachment(n)
		mds = append(mds, md)
	}
	return mds
}

// irMetadataAttachment returns the IR metadata attachment corresponding to
// the given AST metadata attachment.
func (gen *generator) irMetadataAttachment(n ast.MetadataAttachment) *ir.MetadataAttachment {
	panic("not yet implemented")
}
