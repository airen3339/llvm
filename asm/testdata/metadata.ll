declare !foo !0 !bar !{!"baz"} !qux !{!2} void @f1()

define void @f2() !foo !0 !bar !{!"baz"} !qux !{!2} {
	ret void
}

; --- [ Metadata definitions ] -------------------------------------------------

; Empty named metadata definition.
!foo = !{}

; Plain named metadata definition.
!bar = !{!0}

; Multiple metadata IDs.
!baz = !{!0, !1}

; Empty metadata definition.
!0 = !{}

; Plain metadata definition.
!1 = !{!0}

; Multiple metadata IDs.
!2 = !{!0, !1}

; Distinct.
!3 = distinct !{!2}

; Nested metadata.
!4 = !{!{!{!0}}}

; Metadata string.
!5 = !{!"foo"}

; Metadata constant.
!6 = !{i32 42}

; Metadata constant.
!7 = !{!{!"bar"}}
