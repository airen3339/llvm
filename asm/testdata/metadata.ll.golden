declare !bar !{!"baz"} !foo !0 !qux !{!2} void @f1()

define void @f2() !bar !{!"baz"} !foo !0 !qux !{!2} {
; <label>:0
	ret void
}

!foo = !{}

!bar = !{!0}

!baz = !{!0, !1}

!0 = !{}

!1 = !{!0}

!2 = !{!0, !1}

!3 = !{!2}

!4 = !{!{!{!0}}}

!5 = !{!"foo"}

!6 = !{i32 42}

!7 = !{!{!"bar"}}
